language/src/regent/codegen.t

```diff
@@ -2939,13 +2938,45 @@ local function make_partition_projection_functor(cx, expr, loop_index, color_spa
   -- Strip the index for the purpose of checking if this is the
   -- identity projection functor.
   local stripped_index = strip_casts(expr.index)
-  if stripped_index:is(ast.typed.expr.ID) and
-    stripped_index.value == loop_index
+  return stripped_index:is(ast.typed.expr.ID) and stripped_index.value == loop_index
+end
+
+local function is_identity_projection_all(expr, loop_index)
+  if not expr:is(ast.typed.expr.IndexAccess) then
+    return true
+  end
+  return is_identity_projection(expr, loop_index) and
+         is_identity_projection_all(expr.value, loop_index)
+end
+
+local function wrap_partition_internal(node, parent)
+  node.value = ast.typed.expr.Internal {
+    value = values.value(node.value,
+              expr.just(quote end, { impl = parent }),
+              node.value.expr_type),
+    expr_type = node.value.expr_type,
+    annotations = node.annotations,
+    span = node.span
+  }
+  return node
+end
+
+local function make_partition_projection_functor(cx, expr, loop_index, color_space,
+                                                 free_vars, free_vars_setup, requirement)
+  cx = cx:new_local_scope()
+
+  if expr:is(ast.typed.expr.Projection) then
+    expr = expr.region
+  end
+  assert(expr:is(ast.typed.expr.IndexAccess))
+
+  -- Never return 0 for cross products
+  if is_identity_projection_all(expr, loop_index) and
+     not std.is_cross_product(std.as_read(ast.get_base_indexed_node(expr).expr_type))
```

Comment:
Wouldn't it be more correct to return 0 iff the projection functor is exactly one level deep (and is the identity)? If we (hypothetically) supported another nested type in the future, that would avoid a need for an explosion of custom code here.